### DISCUSSION TITLE
Yet another formState Proxy question

### DISCUSSION THREAD

**Discussion by dannyshaw** [react-hook-form/react-hook-form#2806] [Q&A] [answered]
edit: I forgot to fork the CSB so the link here is actually to a working solution

I've got a proxy issue I can't seem to get around.

I'm trying to add a beforeunload handler to my page to prevent the user from losing unsubmitted data.
When I check the isDirty flag in this handler, no matter if I invoke it early it is always false.

I've created a code sandbox here (but see note below):
https://codesandbox.io/s/react-hook-form-onbeforeunload-slrnf?file=/src/index.js

Note, in order to see the issue and not get the beforeunload prompt _from codesandbox_ you'll need to view it in isolation:

1. load https://slrnf.csb.app/
1. open dev dools
1. enter something in the field
1. click the browser home button, breakpoint should trigger with `false` `isDirty` field.

Not sure how to access the up to date value from this handler?

**Comment by bluebill1049** [comment, marked as answer]
> click the browser home button, breakpoint should trigger with false isDirty field.

that's because the useEffect, clean up right?

**Reply by dannyshaw** [reply]
The handler is triggered by attempting to leave the page

**Reply by bluebill1049** [reply]
oh wait, i see it's because `isDirty` is staled, you need to probably pass into a `useCallback` and add `formState` as dep.

**Reply by dannyshaw** [reply]
I'm not sure what you mean. 
The handler is being invoked but the `isDirty` field is incorrect.
Nothing is unmounted, when would the clean up run?

**Reply by dannyshaw** [reply]
haha this isn't working as realtime chat :P

ok i'll experiment with that. thanks

**Reply by dannyshaw** [reply]
I'm still stumped on this one. `useCallback` doesn't seem to have an effect:

forked from above (ie, requires the sandbox in isolation with devtools open):
https://codesandbox.io/s/react-hook-form-onbeforeunload-forked-54c2k?file=/src/index.js
https://54c2k.csb.app/

**Reply by dannyshaw** [reply]
Surely people have implemented onbeforeunload hooks for dirty react-hook-form forms.. 
I'm suspecting maybe I'm approaching this from completely the wrong angle?

**Reply by bluebill1049** [reply]
https://codesandbox.io/s/react-hook-form-onbeforeunload-forked-vmy8d?file=/src/index.js

**Reply by dannyshaw** [reply]
Ok thanks for that, the `useBeforeUnload()` is actually an established hook that I'd rather not modify. 
Here's what I ended up with:

https://codesandbox.io/s/react-hook-form-onbeforeunload-slrnf?file=/src/index.js:768-1069

```javascript
const useCurrentFormState = (formState, getter = (fs) => fs) => {
  const formStateRef = useRef(formState);
  formStateRef.current = formState;
  return () => getter(formStateRef.current);
};

// usage:
const getFormState = useCurrentFormState(formState);

useBeforeunload((e) => {
  const fs = getFormState();
  if (fs.isDirty) {
    e.preventDefault();
  }
});

// or
const getIsDirty = useCurrentFormState(formState, fs => fs.isDirty);
```

**Reply by bluebill1049** [reply]
nice.

**Reply by dannyshaw** [reply]
just spotted a redundancy. pretty much don't need the custom hook now.. lol

**Reply by dannyshaw** [reply]
New final outcome:

```javascript
const formStateRef = useRef(formState);
formStateRef.current = formState;

useBeforeUnload((e) => {
  if (formStateRef.current.isDirty) {
     e.preventDefault();
  }
})
```

**Reply by dannyshaw** [reply]
@bluebill1049 out of interest, am I essentially unproxying here and introducting performance issues?

**Reply by bluebill1049** [reply]
nope.

**Reply by dannyshaw** [reply]
```javascript
const { control } = useForm();

useBeforeUnload((e) => {
  if (control.formStateRef.current.isDirty) {
     e.preventDefault();
  }
})
```

getting lean now!
